plugins/docker/stats_handling.go

```diff
@@ -0,0 +1,361 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import (
+	"strings"
+	"sync"
+	"encoding/json"
+	"time"
+	"fmt"
+
+	"github.com/fsouza/go-dockerclient"
+	. "github.com/mozilla-services/heka/pipeline"
+)
+
+type DockerStat struct {
+	Container		string
+	Time			time.Time
+	StatsString 		string
+	Stat			docker.Stats
+	Fields			map[string]string
+}
+
+type StatsAttachEvent struct {
+	Type string
+	ID   string
+	Name string
+}
+
+type Source struct {
+ 	ID     string
+ 	Name   string
+ 	Filter string
+ 	Types  []string
+ }
+
+type StatsManager struct {
+	sync.RWMutex
+	ir               	InputRunner
+	attached      		map[string]*StatsPump
+	channels      		map[chan *StatsAttachEvent]struct{}
+	client        		DockerClient
+	errors        		chan<- error
+	events        		chan *docker.APIEvents
+	nameFromEnv   		string
+	fieldsFromEnv 		[]string
+	fieldsFromLabels	[]string
+}
+
+func NewStatsManager(endpoint, certPath string, attachErrors chan<- error, nameFromEnv string, fieldsFromEnv []string, fieldsFromLabels []string) (*StatsManager, error) {
+
+	var client DockerClient
+	var err error
+
+	//Get a new DockerClient from attacher
+	client, err = newDockerClient(certPath, endpoint)
+
+	if err != nil {
+		return nil, err
+	}
+
+	m := &StatsManager{
+		attached:         make(map[string]*StatsPump),
+		channels:         make(map[chan *StatsAttachEvent]struct{}),
+		client:           client,
+		errors:           attachErrors,
+		events:           make(chan *docker.APIEvents),
+		nameFromEnv:      nameFromEnv,
+		fieldsFromEnv:    fieldsFromEnv,
+		fieldsFromLabels: fieldsFromLabels,
+	}
+	return m, nil
+}
+
+//The main initial work happens here
+func (m *StatsManager) Run(statsstream chan *DockerStat, closer <-chan struct{}, stopChan chan error) {
+
+	m.attachAll()
+
+	//Launch routine for listening to events
+	go m.handleDockerEvents(stopChan)
+	m.ir.LogMessage("Listen ")
+
+	source := new(Source)
+	events := make(chan *StatsAttachEvent)
+	m.addStatsListener(events)
+	defer m.removeStatsListener(events)
+	//Add listeners for all currently running containers
+	for {
+		select {
+		case event := <-events:
+			m.ir.LogMessage("Event.ID : " + event.ID + " , event.Type : "+ event.Type + " , event.Name : "+  event.Name)
```

Comment:
Small nit, maybe, but I'd rather see the message string set in a value at package scope and fmt.Sprintf used to substitute in the event values.